api/src/main/java/org/openmrs/module/bedmanagement/dao/impl/BedManagementDaoImpl.java

```diff
@@ -299,4 +297,225 @@ public BedLocationMapping getBedLocationMappingByBed(Bed bed) {
         query.setParameter("bed", bed);
         return (BedLocationMapping) query.uniqueResult();
     }
+
+    @Override
+    public List<Bed> getBeds(Integer limit, Integer offset) {
+        String hql = "select b " +
+                "from Bed b " +
+                "where b.voided=:voided";
+
+        Query query = sessionFactory.getCurrentSession().createQuery(hql);
+        query.setParameter("voided", false);
+        if (limit != null) {
+            query.setMaxResults(limit);
+            if (offset != null)
+                query.setFirstResult(offset);
+        }
+
+        return query.list();
+    }
+
+    @Override
+    public List<Bed> getBedsByLocation(Location location, Integer limit, Integer offset) {
+        String hql = "select blm.bed " +
+                "from BedLocationMapping blm " +
+                "join blm.bed b " +
+                "join blm.location l " +
+                "where b.voided=:voided and l=:location";
+
+        Query query = sessionFactory.getCurrentSession().createQuery(hql);
+        query.setParameter("voided", false);
+        query.setParameter("location", location);
+        if (limit != null) {
+            query.setMaxResults(limit);
+            if (offset != null)
+                query.setFirstResult(offset);
+        }
+
+        return query.list();
+    }
+
+    @Override
+    public List<Bed> getBedsByLocationAndBedType(Location location, BedType bedType, Integer limit, Integer offset) {
+        String hql = "select blm.bed " +
+                "from BedLocationMapping blm " +
+                "join blm.bed b " +
+                "join blm.location l " +
+                "where b.voided=:voided and b.bedType=:bedType and l=:location";
+
+        Query query = sessionFactory.getCurrentSession().createQuery(hql);
+        query.setParameter("voided", false);
+        query.setParameter("bedType", bedType);
+        query.setParameter("location", location);
+
+        if (limit != null) {
+            query.setMaxResults(limit);
+            if (offset != null)
+                query.setFirstResult(offset);
+        }
+
+        return query.list();
+    }
+
+    @Override
+    public List<Bed> getBedsByLocationAndStatus(Location location, BedStatus status, Integer limit, Integer offset) {
```

Comment:
Most of your queries are a variant of each other. Depending on the arguments a `join` is added or removed and a condition in the `where` statement is added or removed.
I would suggest to have a util method that can create an Hibernate `Criteria` based on the parameters that are provided or not. A bit on [this model](https://github.com/openmrs/openmrs-core/blob/91aecd7822776ade37b06e906766a139b6b68189/api/src/main/java/org/openmrs/api/db/hibernate/HibernateObsDAO.java#L150) from `HibernateObsDAO`.
If done properly, your DAO layer will have only one resulting public method in the end that can cover all use cases ; and it is the service layer that will offer multiple options while invoking that single method appropriately.
Something along the lines of:
```java
protected Criteria createCriteria(Location location, BedStatus status, BedType bedType, Integer limit, Integer offset) {
  Criteria criteria = ...
    
  if (location != null) {
    // append the criteria accordingly
  }
  if (status != null) {
    // append the criteria accordingly
  }

  ...

  return criteria;
}

public List<Bed> getBeds(Location location, BedStatus status, BedType bedType, Integer limit, Integer offset) {
  return createCriteria(location, status, bedType, limit, offset).list();
}
```

I know the current queries are simple, but please think in terms of **reusability and maintenance**.

Reply:
@mks-d done

Reply:
@mks-d  I have created one generic function that can handle all the possible arguments as you suggested.